### PR TITLE
Limit the amount of dnode metadata in the ARC

### PR DESCRIPTION
**Pull request by dweeezil**
Metadata-intensive workloads can cause the ARC to become permanently
filled with dnode_t objects as they're pinned by the VFS layer.
Subsequent data-intensive workloads may only benefit from about
25% of the potential ARC (arc_c_max - arc_meta_limit).

In order to help track metadata usage more precisely, the other_size
metadata arcstat has replaced with dbuf_size, dnode_size and bonus_size.

The new zfs_arc_dnode_limit tunable, which defaults to 10% of
zfs_arc_meta_limit, defines the minimum number of bytes which is desirable
to be consumed by dnodes.  Attempts to evict non-metadata will trigger
async prune tasks if the space used by dnodes exceeds this limit.

The problem of dnode metadata pinning was observed with the following
testing procedure (in this example, zfs_arc_max is set to 4GiB):

    - Create a large number of small files until arc_meta_used exceeds
      arc_meta_limit (3GiB with default tuning) and arc_prune
      starts increasing.

    - Create a 3GiB file with dd.  Observe arc_mata_used.  It will still
      be around 3GiB.

    - Repeatedly read the 3GiB file and observe arc_meta_limit as before.
      It will continue to stay around 3GiB.

With this modification, space for the 3GiB file is gradually made
available as subsequent demands on the ARC are made.

### CODE REVIEW

**Comment by dweeezil** [comment]
Time to see what the 'bots think of this.  I think this behavior might have started again due to the ca0bf58 (lock contention on arcs_mtx) work but I'm not sure.  I think the commit log message explains things pretty well.  The issue at hand is that pinned dnode metadata can hog the ARC.  This is a particular problem for 0.6.5 users who are setting zfs_arc_max because the 0.6.5-release branch is missing 256fa98 which will frequently cause ```zfs_arc_meta_limit``` to be set to the same value as ```zfs_arc_max``` and allow metadata to 100% clog the ARC.

Issues which may have been caused by or exacerbated by this:
    #4345 
    #4512 (?)
    #4773 
and potentially others in which ```arc_meta_used``` at or above the limit.

This patch effectively gives a lower "priority" to dnode-related metadata in the ARC.

I'm posting this as a PR for buildbot testing and for feedback.

**Comment by dweeezil** [comment]
Rebased to current master.

**Comment by lnicola** [comment]
@dweeezil how do you feel about [this work](https://blogs.oracle.com/roch/entry/rearc)?

**Comment by kernelOfTruth** [comment]
so at least https://github.com/zfsonlinux/zfs/commit/256fa983f46603df3f26c1508ca3a09f6316f36d for inclusion into the next zfs release (for this kind of problem) ?

CC: @nedbass

File: module/zfs/arc.c [hunk 1]
```diff
@@ -7200,4 +7245,7 @@ MODULE_PARM_DESC(zfs_arc_lotsfree_percent,
 module_param(zfs_arc_sys_free, ulong, 0644);
 MODULE_PARM_DESC(zfs_arc_sys_free, "System free memory target size in bytes");
 
+module_param(zfs_arc_dnode_limit, ulong, 0644);
+MODULE_PARM_DESC(zfs_arc_dnode_limit, "Dnode limit for arc size");
```
Comment (behlendorf) [inline module/zfs/arc.c]: Dnode limit for arc size **in bytes**.

Comment (behlendorf) [inline module/zfs/arc.c]: You'll also want to add this to zfs-module-parameters(5)

**Comment by behlendorf** [comment]
@dweeezil I like how this patch splits up ARC_SPACE_OTHER, it gives us much better visibility in to dbuf, dnode, and bonus usage.  It's also minimally invasive which is nice.  One question, how did you end up settling on 10% of the arc_meta_limit being a good default value?

**Comment by tuxoko** [comment]
I think we should hook into drop_inode so we can have better idea of how much inode are not in used and in kernel lru list.

**Comment by dweeezil** [comment]
Getting back to this:

@lnicola I did see the rearc blog entry some time ago and it did sound interesting.  Skimming it again does make me think it's worth further research.

@behlendorf I'll get this cleaned up.  As to the thresholds, I figured we wanted to protect at least _some_ dnode metadata which is where the ```zfs_arc_dnode_limit``` came in.  I wanted to be pretty gentle about this process so it seemed that trying to nibble away at 10% of the overage at a time would be pretty safe.  I'm planning on converting the "nibble ratio" (yes, I'll think of a better name) of 10% (the "/ 10" in the new call to ```arc_prune_async()```) to another tunable.  A better default might be only 5% to cause the eviction of dnode metadata to occur even more gradually.

@tuxoko Yes, as you correctly pointed out, the dentry cache is actually the issue.

@kernelOfTruth In revisiting #4345, It was @logan2211's posting which caught my eye because ```arc_meta_used``` was greater than ```arc_meta_limit``` and was darned near equal to ```size```.  Also his ```other_size``` was a substantial fraction of the ARC.

**Comment by dweeezil** [comment]
If anyone wants to try this on an 0.6.5 release, see the patch in [dweeezil:arc-dnode-limit-0.6.5](https://github.com/dweeezil/zfs/tree/arc-dnode-limit-0.6.5) which also sneaks in the fix from 256fa983.

**Comment by snajpa** [comment]
This is awesome! I've been waiting for something like this. With lots of containers (used as regular VMs) on one node, the IO performance was being hurt by this very problem. If it passes internal testing after merging with tuxoko's ABD based on 0.6.5-stable, I'll be deploying it into production in a few days.

Thanks!

**Comment by behlendorf** [comment]
This LGTM and held up nicely in my local testing.  @snajpa I'll be interested to hear how it holds up to your workloads.

@dweeezil thanks for updating this and adding the documentation.  10% does seem like a reasonable default value and worked well for me in testing.  And since it's a tunable we always adjust it we discover there is some workload where it's a problem.

@lnicola the rearc work is definitely interesting and worth further investigation.  I've often wondered if maintaining the distinction of meta-data vs data is worthwhile.  I think it would be great if somebody worked up a prototype for this change and characterized its impact on a variety of workloads.

**Comment by behlendorf** [comment]
This passed my local testing and looks good.  @dweeezil any last minute concerns or changed before I merge this?

**Comment by dweeezil** [comment]
@behlendorf I think it's good to go.  If you like it, maybe add a sentence to the commit log saying that the previous behavior can be restored by setting ```zfs_arc_dnode_limit``` and ```zfs_arc_meta_limit``` to the same value.

**Comment by behlendorf** [comment]
@dweeezil sounds good, will do.

**Comment by behlendorf** [comment]
Merged as:

25458cb Limit the amount of dnode metadata in the ARC

**Pull request closed without merging**